### PR TITLE
fix(server + client): fix some issues with inferring Record types as procedure outputs

### LINKED ISSUE #4985: bug: Serialize<Record<string, any>> gets inferred as {}
### Provide environment information

trpc commit `43e0da13073c8e5426c4c4fb6c3fb09e598fc6fe`

### Describe the bug

e.g. when using `z.record(z.any())` as output validator, the inferred type on the client is `{}`.

Actual:
```
type Baz = Serialize<Record<string, any>>; // {}
```

Expected:
```
type Baz = Serialize<Record<string, any>>; // Record<string, any>, or { [x: string]: any; }
```

### Link to reproduction

https://youtube.com

### To reproduce

Paste the bug description into `packages/server/src/shared/internal/serialize.ts`

### Additional information

`Serialize<Record<symbol, string>>` also gets inferred as `{ [x: symbol]: string }` which is incorrect. Perhaps that would be a separate issue to file, but I think multiple of these bugs related to `Serialize<Record<...>>` can probably be fixed in one go.

### 👨‍👧‍👦 Contributing

- [X] 🙋‍♂️ Yes, I'd be down to file a PR fixing this bug!

## Funding
* You can sponsor this specific effort via a [Polar.sh](https://polar.sh) pledge below
* We receive the pledge once the issue is completed & verified

<a href="https://polar.sh/trpc/trpc/issues/4985">
<picture>
  <source media="(prefers-color-scheme: dark)" srcset="https://polar.sh/api/github/trpc/trpc/issues/4985/pledge.svg?darkmode=1">
  Fund with Polar
</picture>
</a>

### PR DESCRIPTION
**Pull request by jussisaurio**
…utputs

Closes #4985

## 🎯 Changes

What changes are made in this PR? Is it a feature or a bug fix?

Fixes some issues with inferring Record types as procedure outputs. See linked issue.

## ✅ Checklist

- [x] I have followed the steps listed in the [Contributing guide](https://github.com/trpc/trpc/blob/main/CONTRIBUTING.md).
- [ ] If necessary, I have added documentation related to the changes made.
- [x] I have added or updated the tests related to the changes made.

### CODE REVIEW

File: packages/server/src/shared/internal/serialize.ts [hunk 1]
```diff
@@ -18,18 +18,30 @@ type IsAny<T> = 0 extends T & 1 ? true : false;
...
   T extends [unknown, ...unknown[]] ? SerializeTuple<T> :
   T extends readonly (infer U)[] ? (U extends NonJsonPrimitive ? null : Serialize<U>)[] :
-  T extends object ? Simplify<SerializeObject<UndefinedToOptional<T>>> :
-  never;
+  T extends object ?
+      // Record<K, V> is a special case where we need to ensure that:
+      // 1. we don't match narrower objects like { foo: string } that are also assignable to Record<string, string>
+      // 2. the keys are not symbols
+      // Excluding literal keys (e.g. "foo") from PropertyKey (string | number | symbol) does not exclude anything...
+      Exclude<PropertyKey, keyof T> extends infer TheKeys extends PropertyKey ?
```
Comment (jussisaurio) [inline packages/server/src/shared/internal/serialize.ts:37]: Could probably use something like `type IsRecord<T> = Record<never, never> extends T ? true : false` for this to make the check simpler

**Comment by jussisaurio** [comment]
Turns out records did not have to be specially handled. The problem was that `FilterKeys` in `types.ts` did not handle `any` and `unknown` properly because `any` and `unknown` _always_ match the `Filter` type due to being the broadest types possible. Instead I decided to do the key filtering in `SerializeObject` as a key remapping instead.

File: packages/server/src/shared/internal/serialize.ts [hunk 2]
```diff
@@ -18,18 +18,28 @@ type IsAny<T> = 0 extends T & 1 ? true : false;
...
-  T extends [] ? [] :
-  T extends [unknown, ...unknown[]] ? SerializeTuple<T> :
-  T extends readonly (infer U)[] ? (U extends NonJsonPrimitive ? null : Serialize<U>)[] :
-  T extends object ? Simplify<SerializeObject<UndefinedToOptional<T>>> :
-  never;
+/* prettier-ignore */
+export type Serialize<T> = IsAny<T> extends true
+  ? any
+  : unknown extends T
+  ? unknown
```
Comment (jussisaurio) [inline packages/server/src/shared/internal/serialize.ts:25]: Sorry about this diff. The only thing this commit actually changes about `type Serialize` is it adds explicit handling for `unknown` here

File: packages/server/src/shared/internal/serialize.ts [hunk 3]
```diff
@@ -41,7 +51,20 @@ type SerializeTuple<T extends [unknown, ...unknown[]]> = {
  * @internal
  **/
 export type SerializeObject<T extends object> = {
-  [k in keyof Omit<T, FilterKeys<T, NonJsonPrimitive>>]: Serialize<T[k]>;
+  [k in keyof T as k extends symbol
```
Comment (jussisaurio) [inline packages/server/src/shared/internal/serialize.ts:54]: main beef is here

**Comment by jussisaurio** [comment]
> Turns out records did not have to be specially handled. The problem was that `FilterKeys` in `types.ts` did not handle `any` properly because `any` always matches the `Filter` type. Instead I decided to do the key filtering in `SerializeObject` as a key remapping instead.

`type Foo = any extends "asdfasd" ? true : false;`

What would you intuitively think `Foo` is here? The answer is `boolean` 😂

File: packages/server/src/shared/internal/serialize.ts [hunk 4]
```diff
@@ -36,12 +37,24 @@ type SerializeTuple<T extends [unknown, ...unknown[]]> = {
...
+  // never include entries where the key is a symbol
+  TKey extends symbol ? never : 
+  // always include entries where the value is any
+  IsAny<T[TKey]> extends true ? TKey :
+  // always include entries where the value is unknown
+  unknown extends T[TKey] ? TKey : 
+  // never include entries where the value is a non-JSON primitive
+  T[TKey] extends NonJsonPrimitive ? never : 
+  // otherwise serialize the value
+  TKey;
```
Comment (KATT) [inline packages/server/src/shared/internal/serialize.ts:51]: cc @jussisaurio i made this fn - looks good?

File: packages/server/src/shared/internal/serialize.ts [hunk 5]
```diff
@@ -21,6 +21,7 @@ type JsonReturnable = JsonPrimitive | undefined;
 // prettier-ignore
 export type Serialize<T> =
   IsAny<T> extends true ? any :
+  unknown extends T ? unknown :
```
Comment (KATT) [inline packages/server/src/shared/internal/serialize.ts:24]: fixed the formatting here

File: packages/server/src/shared/internal/serialize.ts (hunk 4 above)
Comment (jussisaurio) [reply packages/server/src/shared/internal/serialize.ts:51]: yeah lgtm

**Review by KATT — APPROVED** [review]

**Pull request merged**